src/red_black_tree.rs

```diff
@@ -331,235 +443,257 @@ impl<
-                    self.transplant(ref_node_index, left);
-                    self.allocator
-                        .clear_register(ref_node_index, Field::Left as u32);
-                    (left, ref_node_index)
-                } else {
-                    let min_right = self.find_min(right);
-                    let min_right_child = self.get_right(min_right);
-                    is_black = self.is_black(min_right);
-                    if min_right == right {
-                        assert!(
-                            min_right_child == SENTINEL
-                                || self.get_parent(min_right_child) == right
-                        );
-                    } else {
-                        self.transplant(min_right, min_right_child);
-                        self.connect(min_right, right, Field::Right as u32);
-                    }
-                    self.transplant(ref_node_index, min_right);
-                    self.connect(min_right, left, Field::Left as u32);
-                    if self.is_red(ref_node_index) {
-                        self.color_red(min_right)
-                    } else {
-                        self.color_black(min_right)
-                    }
-                    self.allocator
-                        .clear_register(ref_node_index, Field::Left as u32);
-                    self.allocator
-                        .clear_register(ref_node_index, Field::Right as u32);
-                    (min_right_child, ref_node_index)
-                };
-                self.allocator
-                    .clear_register(ref_node_index, Field::Parent as u32);
-                self.allocator.clear_register(delete_node_index, COLOR);
-                self.allocator.remove_node(delete_node_index);
-                if is_black && self.fix_remove(pivot_node_index) == None {
-                    return None;
+            let RBNode {
+                key: curr_key,
+                value: curr_value,
+            } = *self.allocator.get(curr_node_index).get_value();
+            let target = match key.cmp(&curr_key) {
+                Ordering::Less => self.get_left(curr_node_index),
+                Ordering::Greater => self.get_right(curr_node_index),
+                Ordering::Equal => {
+                    self._remove_tree_node(curr_node_index);
+                    return Some(curr_value);
                 }
-                return Some(ref_value);
             };
             if target == SENTINEL {
                 return None;
             }
-            ref_node_index = target
+            curr_node_index = target
+        }
+    }
+
+    fn _remove_tree_node(&mut self, node_index: u32) {
+        let mut is_black = self.is_black(node_index);
+        let left = self.get_left(node_index);
+        let right = self.get_right(node_index);
+        let (pivot_node_index, parent_and_dir) = if self.is_leaf(node_index) {
+            if !self.is_root(node_index) {
```

Comment:
Issue (2b) Add an entirely new case to handle when the node_index is a leaf node as well when we are removing the final node in the tree